pkg/storage/tsdb/tombstones.go

```diff
@@ -0,0 +1,342 @@
+
+	// Check if the tombstone already exists. Could be the case the same request was made
+	// and is already in the middle of deleting series. Creating a new tombstone would restart
+	// the progress
+	tombstoneExists, err := TombstoneExists(ctx, userBkt, userID, tombstone.RequestID, tombstone.State)
+	if err != nil {
+		level.Error(userLogger).Log("msg", "unable to check if the same tombstone already exists", "requestID", tombstone.RequestID, "err", err)
+	} else if tombstoneExists {
+		return ErrTombstoneAlreadyExists
+	}
+
+	return errors.Wrap(userBkt.Upload(ctx, fullTombstonePath, bytes.NewReader(data)), "upload tombstone file")
+}
+
+func TombstoneExists(ctx context.Context, bkt objstore.BucketReader, userID string, requestID string, state BlockDeleteRequestState) (bool, error) {
+	// Add the state of the request using the filename extension
+	filename := requestID + "." + string(state) + ".json"
+	fullTombstonePath := path.Join(TombstonePath, filename)
+
+	exists, err := bkt.Exists(ctx, fullTombstonePath)
+
+	if exists || err != nil {
+		return exists, err
+	}
+
+	return false, nil
+}
+
+func GetDeleteRequestByIdForUser(ctx context.Context, bkt objstore.Bucket, cfgProvider bucket.TenantConfigProvider, userID string, requestId string) (*Tombstone, error) {
+	userLogger := util_log.WithUserID(userID, util_log.Logger)
+	userBucket := bucket.NewUserBucketClient(userID, bkt, cfgProvider)
+
+	found := []*Tombstone{}
+
+	for _, state := range AllDeletionStates {
+		filename := requestId + "." + string(state) + ".json"
+
+		exists, err := TombstoneExists(ctx, userBucket, userID, requestId, state)
+		if err != nil {
+			return nil, err
+		}
+
+		if exists {
+			t, err := readTombstoneFile(ctx, userBucket, userID, path.Join(TombstonePath, filename))
+			if err != nil {
+				return nil, err
+			}
+			found = append(found, t)
+		}
+	}
+
+	if len(found) == 0 {
+		return nil, nil
+	}
+
+	// If there are multiple tombstones with the same request id but different state, should delete the files with the lower state
+	for _, ts := range found[0 : len(found)-1] {
+		level.Info(userLogger).Log("msg", "Found extra tombstone file with outdated state. Will delete it.", "requestID", ts.RequestID, "state", ts.State)
+		if err := DeleteTombstoneFile(ctx, bkt, cfgProvider, ts); err != nil {
+			level.Error(userLogger).Log("msg", "Unable to delete redundant tombstone file.", "request_id", ts.RequestID, "state", ts.State, "err", err)
+		}
+	}
+
+	return found[len(found)-1], nil
+
+}
+
+func GetAllDeleteRequestsForUser(ctx context.Context, bkt objstore.Bucket, cfgProvider bucket.TenantConfigProvider, userID string) ([]*Tombstone, error) {
+	userBucket := bucket.NewUserBucketClient(userID, bkt, cfgProvider)
+
+	// add all the tombstones to a map and check for duplicates,
+	// if a key exists with the same request ID (but two different states)
+	tombstoneMap := make(map[string]*Tombstone)
+	err := userBucket.Iter(ctx, "tombstones/", func(s string) error {
```

Comment:
I think this too can be simplified as follows:
- Initialize the tombstone map
- Sort all the tombstones in the bucket by state
- Iterate over the sorted tombstones. set `tombstonesMap[id] =  tombstoneFile`
- What's left in the map would be the most recent state.

Reply:
As per your other comment, I removed the `removeDuplicateTombstone` call from the get functions and add it during the cleanup in the compactor. This should simplify it a bit. 

I think with the suggestion you made, it might be less efficient if I interpreted it correctly: 

- Iterate over the tombstones in the bucket and add them all to a slice. 
- Sort the slice
-  Iterate over the slice and add it to the map
- iterate over the map and add it to a new slice
- return the slice

This will add another pass over the slice and a sort operation. 

Let me know if I misinterpreted your suggestion.